SOURCES/etc/xapi.d/plugins/updater.py

```diff
@@ -75,6 +78,38 @@ def upgrade():
     pass
 
 
+# returns a JSON dict {repo_id: proxy}
+def get_proxies(session, args):
+    config = ConfigParser.ConfigParser({'proxy': '_none_'})
+    config.read('/etc/yum.repos.d/xcp-ng.repo')
+    new_dict = dict((section, config.get(section, 'proxy', '_none_')) for section in config.sections())
+    return json.dumps(new_dict)
+
+
+# expects a JSON dict in a 'proxies' argument. The dict should be of the form {repo_id: proxy} with the special proxy
+# '_none_' used for removal.
+# example: {"xcp-ng-base": "http://192.168.100.82:3142"}
+def set_proxies(session, args):
+    # '{"xcp-ng-base": "http://192.168.100.82:3142"}'
+    # '{"xcp-ng-base": "_none_"}'
+    special_url_prefix = 'http://HTTPS///'
+    https_url_prefix = 'https://'
+    proxies = json.loads(args['proxies'])
+    config = ConfigParser.ConfigParser()
+    config.read('/etc/yum.repos.d/xcp-ng.repo')
+    for section in proxies:
+        if config.has_section(section):
+            config.set(section, 'proxy', proxies[section])
+            url = config.get(section, 'baseurl')
+            if proxies[section] == '_none_' and url.startswith(special_url_prefix):
+                config.set(section, 'baseurl', https_url_prefix + url[len(special_url_prefix):])
+            elif proxies[section] != '_none_' and url.startswith(https_url_prefix):
+                config.set(section, 'baseurl', special_url_prefix + url[len(https_url_prefix):])
+    with open('/etc/yum.repos.d/xcp-ng.repo', 'wb') as configfile:
```

Comment:
Same comment about using a constant.

Let's also create a backup of the file as `/etc/yum.repos.d/xcp-ng.repo.xosave` ?

Reply:
I am a bit torn, because if we back up, should we keep more than one revision and rotate them?

Reply:
Just one seems enough to me, not to be used by the plugin directly, only as a backup just in case

Reply:
my vision is that someone clicks in the UI, gets an error, click again to try to force, and there is no backup.

Reply:
That's why we shouldn't write the file at all where there's an error, isn't it?

Reply:
Yeah, I was avoiding that because I have no good way to report an error. In particular, the `else` would catch a perfectly legit 'http://' url. And if there is an error I have to show the user what I was about to do, what I was expected to find in the file, color the difference in red, and well now that you have all that in JS, the python is mostly an hindrance.

Reply:
The plugin is supposed to work with an unaltered xcp-ng.repo, but it's fine and even good if it refuses to do the changes if it meets an unexpected http:// URL or when we're asking to modify a repo that doesn't exist in the file. Simple error messages or error codes would be sufficient.

It would be good to try/except the file reads and writes too. Anything goes wrong? Just return an error, preferrably different error codes or messages for different errors. It's OK if the user doesn't understand the error if either we or someone reading the code can.

Reply:
What about letting the plugin handle the exceptions?

Reply:
I don't know exactly how those plugins work. Do you have an example situation and how it would be handled?